benchexec/tools/template.py

```diff
@@ -22,17 +35,386 @@ class UnsupportedFeatureException(benchexec.BenchExecException):
+
+        @classmethod
+        def with_files(cls, input_files, *, property_file=None, options=None):
+            return cls(
+                input_files=input_files,
+                identifier=None,
+                property_file=property_file,
+                options=options,
+            )
+
+        @classmethod
+        def without_files(cls, identifier, *, property_file=None, options=None):
+            return cls(
+                input_files=[],
+                identifier=identifier,
+                property_file=property_file,
+                options=options,
+            )
+
+        @property
+        def input_files(self):
+            self.require_input_files()
+            return self.input_files_or_empty
+
+        @property
+        def input_files_or_identifier(self):
+            """
+            Return either the sequence of input files or a one-element sequence with the
+            identifier. Useful for adding either to the command line arguments.
+            """
+            return self.input_files_or_empty or tuple(self.identifier)
+
+        def require_input_files(self):
+            """
+            Check that there is at least one path in input_files and raise appropriate
+            exception otherwise
+            """
+            if not self.input_files_or_empty:
+                raise UnsupportedFeatureException(
+                    "Tool does not support tasks without input files"
+                )
+
+    class ResourceLimits(
+        namedtuple(
+            "ResourceLimits",
+            ["cputime", "cputime_hard", "walltime", "memory", "cpu_cores"],
+        )
+    ):
+        """
+        Represent resource limits of a run. While this class is technically a tuple,
+        this should be seen as an implementation detail and the order of elements in the
+        tuple should not be considered. New fields may be added in the future.
+        Each field contains a positive int or None, which means no limit.
+
+        Explanation of fields:
+        cputime: CPU-time limit in seconds after which the tool will receive
+            a termination signal and the result will be counted as timeout
+        cputime_hard: CPU-time limit in seconds after which the tool will be killed
+            forcibly (always greater or equal than cputime)
+        walltime: Wall-time limit in seconds after which the tool will be killed
+        memory: Memory limit in bytes
+        cpu_cores: Number of CPU cores allowed to be used
+
+        The CPU-time limits will either both have a value of both be None.
+        """
+
+        def __new__(
+            cls,
+            cputime=None,
+            cputime_hard=None,
+            walltime=None,
+            memory=None,
+            cpu_cores=None,
+        ):
+            return super().__new__(
+                cls, cputime, cputime_hard, walltime, memory, cpu_cores
+            )
+
+        pass
+
+
+class BaseTool(object):
```

Comment:
can be simplified to `class BaseTool:`

Reply:
I prefer to not change the code of the old code to keep the diff minimal.

Reply:
I agree - I only mentioned it because it already is part of the git diff. But I get your point.